python/ray/autoscaler/v2/instance_manager/node_provider.py

```diff
@@ -1,16 +1,175 @@
+    cluster. It provides the following main functionalities:
+        - Update the cloud instances:
+            1. Given a target running set of cloud instances, may launch new ones.
+            2. Terminate existing running instances.
+        - Get the running cloud instances in the cluster.
+        - Poll the errors that happened for the updates to the cloud instance provider.
+
+    Below properties of the cloud instance provider are assumed with this interface:
+
+    1. Eventually consistent
+    The cloud instance provider is expected to be eventually consistent with the
+    cluster state. For example, when a cloud instance is request to be terminated
+    ,the provider may not immediately reflect the change in its state.
+    However, the provider is expected to eventually reflect the change in its state.
+
+    2. Asynchronous
+    The provider could also be asynchronous, where the termination/launch
+    request may not immediately return the result of the request.
+
+    3. Unique cloud instance ids
+    Cloud instance ids are expected to be unique across the cluster.
+
+    4. Idempotent updates
+    For the update APIs (e.g. update_cluster, terminate), the provider may use the
+    update ids to provide idempotency.
+
+    Usage:
+        ```
+            provider: ICloudInstanceProvider = ...
+
+            # Update the cluster with a designed shape.
+            provider.update(
+                target_running_nodes={
+                    "worker_nodes": 10,
+                    "ray_head": 1,
+                },
+                to_terminate=["node_1", "node_2"],
+                id="update_1",
+            )
+
+            # Poll the state of the cloud instance provider.
+            running = provider.get_running()
+
+            # Poll the errors
+            errors = provider.poll_errors()
+
+            # Process the state of the provider.
+            ...
+        ```
+    """
+
+    @abstractmethod
+    def get_running(self) -> Dict[CloudInstanceId, CloudInstance]:
+        """Get the running cloud instances in the cluster.
+
+        Returns:
+            A dictionary of the running cloud instances in the cluster.
+            The key is the cloud instance id, and the value is the cloud instance.
+        """
+        pass
+
+    @abstractmethod
+    def update(
+        self,
+        target_running_nodes: Dict[NodeType, int],
+        to_terminate: List[CloudInstanceId],
+        id: str,
+    ) -> None:
+        """Update the provider target cluster shape.
+
+        The provider should launch new nodes to match the target shape.
+
+        Args:
+            target_running_nodes: the target cluster shape (number of running nodes by
+                type).
+            to_terminate: the nodes to terminate.
+            id: the unique id of the update request.
+        """
+        pass
+
+    @abstractmethod
+    def poll_errors(self) -> List[CloudInstanceProviderError]:
```

Comment:
is this an internal method? do we need this to be public?

Reply:
this will need to be external method. This is used for instance manager to retrieve errors happened if there's any in the node provider. 

This is needed since the node provider doesn't return anything upon the `update` request.

Reply:
what's the semantic of this method? i.e. what happens if i query the api twice?

Reply:
errors happen would be retrieved exactly once, i.e. a second call would be empty if there's no new errors occurred after the first call.  It's like a queue (which is probably how this will be implemented). 

Also, I should proabbly mention this call is not thread-safe (although it could be made to but I don't think it needs for now).